### PR TITLE
Make sure openmp include dir is added to compiler include dirs

### LINKED ISSUE #18478: 4.1.1 (and master) does not compile on macOS with WITH_OPENMP
##### System information (version)

- macos 10.15.7
- OpenMP installed with `brew install libomp` and headers and libraries present in `/usr/local/`
- Compiler Apple clang 12.0.0

##### Detailed description

Using this build configuration:
```
install_opencv() {
    (
        local version=4.1.1
        local archive_name=opencv-$version.tar.gz
            wget -N https://github.com/opencv/opencv/archive/$version.tar.gz --output-document $archive_name && \
            tar -v -xzf $archive_name && \
            cd opencv-$version && \
            mkdir -p build && \
            cd build && \
            cmake $CMAKE_BUILD_FLAGS -DBUILD_LIST=core,imgproc,highgui,calib3d .. \
            -DENABLE_PRECOMPILED_HEADERS=OFF\
            -DBUILD_features2d=OFF\
            -DBUILD_gapi=OFF\
            -DBUILD_java_bindings_gen=OFF\
            -DBUILD_ml=OFF\
            -DBUILD_objdetect=OFF\
            -DBUILD_photo=OFF\
            -DBUILD_video=OFF\
            -DBUILD_videoio=OFF\
            -DBUILD_opencv_python_bindings_generator=OFF\
            -DBUILD_ZLIB=OFF\
            -DBUILD_TIFF=OFF\
            -DBUILD_JASPER=OFF\
            -DBUILD_JPEG=ON\
            -DBUILD_PNG=OFF\
            -DBUILD_OPENEXR=OFF\
            -DBUILD_WEBP=OFF\
            -DBUILD_IPP_IW=OFF\
            -DBUILD_ITT=OFF\
            -DWITH_AVFOUNDATION=OFF\
            -DWITH_CUDA=OFF\
            -DWITH_EIGEN=ON\
            -DWITH_FFMPEG=OFF\
            -DWITH_GSTREAMER=OFF\
            -DWITH_VTK=OFF\
            -DWITH_GTK=ON\
            -DWITH_GTK_2_X=OFF\
            -DWITH_IPP=OFF\
            -DWITH_JASPER=OFF\
            -DWITH_JPEG=ON\
            -DWITH_WEBP=OFF\
            -DWITH_OPENMP=ON\	# use openmp
            -DMKL_WITH_OPENMP=ON\
            -DWITH_OPENEXR=OFF\
            -DWITH_OPENNI=OFF\
            -DWITH_OPENNI2=OFF\
            -DWITH_TIFF=OFF\
            -DWITH_V4L=OFF\
            -DWITH_OPENCLAMDFFT=OFF\
            -DWITH_OPENCLAMDBLAS=OFF\
            -DWITH_PROTOBUF=OFF\
            -DWITH_IMGCODEC_HDR=OFF\
            -DWITH_IMGCODEC_SUNRASTER=OFF\
            -DWITH_IMGCODEC_PFM=OFF\
            -DWITH_QUIRC=OFF\
            -DBUILD_opencv_apps=OFF\
            -DBUILD_opencv_js=OFF\
            -DBUILD_PERF_TESTS=OFF\
            -DBUILD_TESTS=OFF\
            -DBUILD_JAVA=OFF\
            -DBUILD_opencv_python2=OFF\
            -DBUILD_opencv_python3=OFF && \
            make -j12 install
    )
}
```

Fails to compile since `cmake/OpenCVFindFrameworks.cmake` uses such code:
```
# --- OpenMP ---
if(WITH_OPENMP)
  find_package(OpenMP)
  if(OPENMP_FOUND)
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
	# missing OpenMP include and/or library dirs like so:
    # link_libraries(OpenMP::OpenMP_CXX)
  endif()
  set(HAVE_OPENMP "${OPENMP_FOUND}")
endif()
```

which does not bring `-I/usr/local/include`, `-L/usr/local/lib` and `-lomp` into the compile commands, only the flags `-Xclang -fopenmp`. Even if that were added here, I don't think it fixes the issue for the 3rdparty cmake projects.

You get something like this:
```
…
[  2%] Building C object 3rdparty/libjasper/CMakeFiles/libjasper.dir/jas_cm.c.o
[  2%] Building C object 3rdparty/libjpeg-turbo/CMakeFiles/libjpeg-turbo.dir/src/jcapimin.c.o
[  2%] Building C object 3rdparty/libwebp/CMakeFiles/libwebp.dir/src/dec/alpha_dec.c.o
[  2%] Building C object 3rdparty/zlib/CMakeFiles/zlib.dir/crc32.c.o
/Users/rasmus/Downloads/opencv/build/3rdparty/ippicv/ippicv_mac/iw/src/iw_core.c:44:18: fatal error: 'omp.h' file not found
        #include <omp.h> // Use OMP atomics
                 ^~~~~~~
```

##### Steps to reproduce

See bash function above.

Installed brew packages before opencv build:
```
BREW_PACKAGES=(
    cmake
    pkg-config
    doxygen
    zmq
    yaml-cpp
    eigen
    clang-format
    vtk@8.2
    tclap
    libomp
    libusb
    ffmpeg
    wget
    curl
    qt5
    tbb
    xsimd
    xtensor
    automake
    autoconf
    flann
)
brew install ${BREW_PACKAGES[*]}
brew upgrade ${BREW_PACKAGES[*]}
brew link --force vtk@8.2
```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
 - [x] I updated to latest OpenCV version and the issue is still there

### PR DESCRIPTION
**Pull request by themightyoarfish**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

Fixes #18478

### CODE REVIEW

**Comment by themightyoarfish** [comment]
I'm not sure if I added the `OpenMP_CXX_LIBRARIES` in the right place. This variable doesn't seem to be required everywhere, but it is on macos, linking to the `OpenMP::OpenMP_CXX` target brings in compiler options I think, but apparently not the `limomp` runtime library, not on macos anyways.

**Comment by themightyoarfish** [comment]
Guess not; adding the libraries like this does not make client projects link. It's weird this doesn't already show up when linking opencv itself.

**Comment by themightyoarfish** [comment]
`modules/core/misc/plugins/parallel_openmp/CMakeLists.txt` seemingly also unused.
OpenMP gets enabled apparently anyhow, though the process is mysterious to me.

**Comment by themightyoarfish** [comment]
I find no place to link to `OpenMP_CXX_LIBRARIES` that would actually make its contents (`/usr/local/lib/libomp.dylib` for me)  appear in the verbose make output, so it seems like it's never getting linked.

File: cmake/OpenCVFindFrameworks.cmake [hunk 1]
```diff
@@ -32,6 +32,7 @@ if(WITH_OPENMP)
   if(OPENMP_FOUND)
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+    ocv_include_directories(${OpenMP_CXX_INCLUDE_DIRS})
```
Comment (alalek) [inline cmake/OpenCVFindFrameworks.cmake:35]: Old CMake doesn't know about these variables. Need to add checks to properly handle that.

Ref: https://cmake.org/cmake/help/v3.5/module/FindOpenMP.html

File: modules/core/CMakeLists.txt [hunk 2]
```diff
@@ -167,6 +167,10 @@ if(HAVE_HPX)
   ocv_target_link_libraries(${the_module} LINK_PRIVATE "${HPX_LIBRARIES}")
 endif()
 
+if(HAVE_OPENMP)
+    ocv_target_link_libraries(${the_module} PUBLIC ${OpenMP_CXX_LIBRARIES} OpenMP::OpenMP_CXX)
```
Comment (alalek) [inline modules/core/CMakeLists.txt:171]: Using both approaches looks error prone.

File: cmake/OpenCVFindFrameworks.cmake (hunk 1 above)
Comment (themightyoarfish) [reply cmake/OpenCVFindFrameworks.cmake:35]: Will the variable not simply be empty in that case?

File: modules/core/CMakeLists.txt (hunk 2 above)
Comment (themightyoarfish) [reply modules/core/CMakeLists.txt:171]: You mean doing this here and in `OpenCVFrameworks.cmake`. Might be the latter can be omitted; i was just looking for a place to add the link_libraries call. There wasn't any in the `OpenCVFrameworks.cmake` file and I have no idea about the proper place. Where do you suggest to put this?

Comment (themightyoarfish) [reply modules/core/CMakeLists.txt:171]: And I guess here there would be a problem if the library target does not exist in older cmake versions. I assume an `if(TARGET…`  would be sufficient?

File: cmake/OpenCVFindFrameworks.cmake (hunk 1 above)
Comment (themightyoarfish) [reply cmake/OpenCVFindFrameworks.cmake:35]: any advice on this?

**Comment by asmorkalov** [comment]
@themightyoarfish Friendly reminder.

**Comment by themightyoarfish** [comment]
Awaiting feedback. 

On May 20, 2022 9:49:14 AM GMT+02:00, Alexander Smorkalov ***@***.***> wrote:
***@***.*** Friendly reminder.

File: modules/core/CMakeLists.txt (hunk 2 above)
Comment (alalek) [reply modules/core/CMakeLists.txt:171]: Yes, this should work:

```
if(TARGET OpenMP::OpenMP_CXX)
  ... use modern approach ...
else()
  ... use legacy approach (keep existed behavior the same) ...
endif()
```

Comment (themightyoarfish) [reply modules/core/CMakeLists.txt]: fixed

**Comment by themightyoarfish** [comment]
I've update this to hopefully work with old cmake as well. `OpenMP_CXX_INCLUDE_DIRS` will simply be empty, which is ok, and I am checking for the `OpenMP::` library target to exist before linking to it.

**Comment by alalek** [comment]
superseded by #23095

**Pull request closed without merging**